### PR TITLE
CRPS improvements: fast and unbiased kernel implementations

### PR DESCRIPTION
**Pull request by nbren12**
# Modulus Pull Request

## Description

Implements the unbiased (fair) versions of the CRPS formula.

This also harmonizes the CRPS implementations. Following Zamo and Naveau, the "kernel" and "sort" methods give identical answers. Also, there is a fast sort-based algorithm for the kernel formula (see the derivation in `_kernel_crps_implementation`). This is also used [by weather bench](https://github.com/google-research/weatherbench2/blob/main/weatherbench2/metrics.py#L715).

I could leave the existing `kcrps` and `_crps_from_empirical_crps` formulas, but seems worthwhile to harmonize the implementations.

Zamo, M., & Naveau, P. (2018). Estimation of the Continuous Ranked Probability Score with Limited Information and Applications to Ensemble Weather Forecasts. Mathematical Geosciences, 50(2), 209–234. https://doi.org/10.1007/s11004-017-9709-7

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: modulus/metrics/general/crps.py [hunk 1]
```diff
@@ -51,23 +75,41 @@ def kcrps(pred: Tensor, obs: Tensor, dim: int = 0):
...
+
+    Notes
+    -----
+    Zamo, M., & Naveau, P. (2018). Estimation of the Continuous Ranked
+    Probability Score with Limited Information and Applications to Ensemble
+    Weather Forecasts. Mathematical Geosciences, 50(2), 209–234.
+    https://doi.org/10.1007/s11004-017-9709-7
+    """
+    pred = pred.transpose(dim, -1)
+    return _kernel_crps_implementation(pred, obs, unbiased=True)
```
Comment (dallasfoster) [inline modulus/metrics/general/crps.py:112]: Do you feel strongly about having both a `fair_crps` and `kcrps`? It seems to me like they could be condensed into a single function with a `biased` as a keyword argument.

File: modulus/metrics/general/crps.py [hunk 2]
```diff
@@ -51,23 +75,41 @@ def kcrps(pred: Tensor, obs: Tensor, dim: int = 0):
...
+
+
+    Notes
+    -----
+    Zamo, M., & Naveau, P. (2018). Estimation of the Continuous Ranked
+    Probability Score with Limited Information and Applications to Ensemble
+    Weather Forecasts. Mathematical Geosciences, 50(2), 209–234.
+    https://doi.org/10.1007/s11004-017-9709-7
+    """
+    pred = pred.transpose(dim, -1)
```
Comment (dallasfoster) [inline modulus/metrics/general/crps.py:111]: This, in addition to similar code above, needs to be replaced with
```
pred = pred.unsqueeze(-1).transpose(dim, -1).squeeze(dim).contiguous()
```

File: test/metrics/test_metrics_general.py [hunk 3]
```diff
@@ -226,6 +226,60 @@ def test_histogram(device, input_shape, rtol: float = 1e-3, atol: float = 1e-3):
...
+
+@pytest.mark.parametrize("device", ["cuda:0", "cpu"])
+def test_fair_crps_dim_arg_works(device):
+    pred = torch.randn((10, 100), device=device)
+
+    value = crps.fair_crps(pred, torch.zeros([pred.size(0)], device=device), dim=1)
+    assert value.shape == (10,)
+
+    value = crps.fair_crps(pred, torch.zeros([pred.size(1)], device=device), dim=0)
+    assert value.shape == (100,)
```
Comment (dallasfoster) [inline test/metrics/test_metrics_general.py:280]: Try with the following:

```
pred = torch.randn((5, 10, 100), device = device)
```

**Review by dallasfoster — CHANGES_REQUESTED** [review]

File: modulus/metrics/general/crps.py (hunk 1 above)
Comment (nbren12) [reply modulus/metrics/general/crps.py:112]: done.

File: modulus/metrics/general/crps.py (hunk 2 above)
Comment (nbren12) [reply modulus/metrics/general/crps.py:111]: good catch. I replaced with [`torch.movedim`](https://pytorch.org/docs/stable/generated/torch.movedim.html#torch.movedim).

File: test/metrics/test_metrics_general.py (hunk 3 above)
Comment (nbren12) [reply test/metrics/test_metrics_general.py]: done. nice suggestion to catch [the bug you found above](https://github.com/NVIDIA/modulus/pull/577#discussion_r1671263418).

**Comment by nbren12** [comment]
@dallasfoster Will the existing tests catch an error in the numerical value of kcrps---as opposed to its asymptotic value for large ensemble size? I want to ensure this hasn't change, so may add a regression test if there isn't one already.

File: modulus/metrics/general/crps.py (hunk 2 above)
Comment (dallasfoster) [reply modulus/metrics/general/crps.py]: Great, thanks!

File: test/metrics/test_metrics_general.py (hunk 3 above)
Comment (dallasfoster) [reply test/metrics/test_metrics_general.py]: Test looks great.

**Comment by dallasfoster** [comment]
Thanks @nbren12, I added a regression and another finite ensemble accuracy test. LGTM.

**Review by dallasfoster — APPROVED** [review]

**Review by NickGeneva — APPROVED** [review]
LGTM, thanks for the derivation notes
